controllers/scoring.py

```diff
@@ -216,8 +219,36 @@ def get_cone_pickup(self) -> tuple[Pose2d, Rotation2d]:
         )
 
     def get_score_location(self) -> tuple[tuple[Pose2d, Rotation2d], Setpoint]:
-        # TODO: pick actual scoring node
-        node_trans3d = BLUE_NODES[int(self.DESIRED_ROW)][int(self.DESIRED_COLUMN)]
+        if self.snap_to_closest_node:
+            score_locations = []
+            if self.get_current_piece() == GamePiece.CONE:
+                score_locations = [
+                    self.score_location_from_node(Node(Rows.HIGH, i), self.is_red())
+                    for i in range(9)
+                    if i % 3 != 1
+                ]
+            elif self.get_current_piece == GamePiece.CUBE:
+                score_locations = [
+                    self.score_location_from_node(Node(Rows.HIGH, i), self.is_red())
+                    for i in range(9)
+                    if i % 3 == 1
+                ]
+            score_poses = [t[0][0] for t in score_locations]
+            if len(score_poses) == 0:
+                return self.score_location_from_node(
+                    Node(Rows(self.DESIRED_ROW), self.DESIRED_COLUMN), self.is_red()
+                )
+            nearest = self.movement.chassis.get_pose().nearest(score_poses)
+            return next(loc for loc in score_locations if loc[0][0] == nearest)
+        else:
+            return self.score_location_from_node(
+                Node(Rows(self.DESIRED_ROW), self.DESIRED_COLUMN), self.is_red()
+            )
+
+    def score_location_from_node(
+        self, node: Node, is_red: bool
+    ) -> tuple[tuple[Pose2d, Rotation2d], Setpoint]:
+        node_trans3d = BLUE_NODES[int(node.row.value)][int(node.col)]
```

Comment:
isn't the row value already an int?

```suggestion
        node_trans3d = BLUE_NODES[node.row.value][int(node.col)]
```